zhaquirks/tuya/builder/__init__.py

```diff
@@ -0,0 +1,493 @@
+            attribute_name=attribute_name,
+            type=t.Bool,
+            access=foundation.ZCLAttributeAccess.Read
+            | foundation.ZCLAttributeAccess.Report,
+        )
+        self.binary_sensor(
+            attribute_name=attribute_name,
+            cluster_id=TUYA_CLUSTER_ID,
+            endpoint_id=endpoint_id,
+            device_class=device_class,
+            initially_disabled=initially_disabled,
+            attribute_initialized_from_cache=attribute_initialized_from_cache,
+            translation_key=translation_key,
+            fallback_name=fallback_name,
+        )
+
+        return self
+
+    def tuya_sensor(
+        self,
+        dp_id: int,
+        attribute_name: str,
+        type: type,
+        converter: Optional[Callable[[Any,], Any,]] = None,
+        dp_converter: Optional[Callable[[Any,], Any,]] = None,
+        endpoint_id: int = 1,
+        divisor: int = 1,
+        multiplier: int = 1,
+        entity_type: EntityType = EntityType.STANDARD,
+        device_class: SensorDeviceClass | None = None,
+        state_class: SensorStateClass | None = None,
+        unit: str | None = None,
+        initially_disabled: bool = False,
+        attribute_initialized_from_cache: bool = True,
+        translation_key: str | None = None,
+        fallback_name: str | None = None,
+    ) -> QuirkBuilder:  # fmt: skip
+        """Add an EntityMetadata containing ZCLSensorMetadata and return self.
+
+        This method allows exposing a sensor entity in Home Assistant.
+        """
+
+        self.tuya_dp_attribute(
+            dp_id=dp_id,
+            attribute_name=attribute_name,
+            type=type,
+            converter=converter,
+            dp_converter=dp_converter,
+            access=foundation.ZCLAttributeAccess.Read
+            | foundation.ZCLAttributeAccess.Report,
+        )
+        self.sensor(
+            attribute_name=attribute_name,
+            cluster_id=TUYA_CLUSTER_ID,
+            endpoint_id=endpoint_id,
+            divisor=divisor,
+            multiplier=multiplier,
+            entity_type=entity_type,
+            device_class=device_class,
+            state_class=state_class,
+            unit=unit,
+            initially_disabled=initially_disabled,
+            attribute_initialized_from_cache=attribute_initialized_from_cache,
+            translation_key=translation_key,
+            fallback_name=fallback_name,
+        )
+
+        return self
+
+    def add_to_registry(self) -> QuirksV2RegistryEntry:
+        """Build the quirks v2 registry entry."""
+
+        attr_dict: dict[int, foundation.ZCLAttributeDef] = self.new_attributes
+
+        class TuyaReplacementCluster(TuyaMCUCluster):
+            """Replacement Tuya Cluster."""
+
+            data_point_handlers: dict[int, str]
+            dp_to_attribute: dict[int, DPToAttributeMapping]
+            attributes = TuyaMCUCluster.attributes.copy()
+            attributes.update(attr_dict)
```

Comment:
Assuming the `TuyaMCUCluster` and underlying clusters finally get updated to use `AttributeDefs`, this won't work anymore.
As far as I can see, `TuyaMCUCluster` (and the underlying clusters) do not actually have any attributes defined, only server and client commands that need to be updated.

Ideally, we'd already use the new zigpy-style `AttributeDefs` here. It might not be as clean though..

Just for reference, here's the "programmatic creation and populating" of `AttributeDefs` in zigpy:
https://github.com/zigpy/zigpy/blob/dfeb958b1f737b0a6c9dedbdb7fc169a188a516e/zigpy/zcl/__init__.py#L135-L143
(and the whole conversion from old-style to new-style attributes definition:)
https://github.com/zigpy/zigpy/blob/dfeb958b1f737b0a6c9dedbdb7fc169a188a516e/zigpy/zcl/__init__.py#L95-L211

Reply:
If you don't want to change this at the moment, we could also leave it like this for a first version, but we don't really want to add "old-style zigpy attributes" at this point, though this is a somewhat different case compared to most other quirks.